### ISSUE TITLE
Deploying without `mike deploy`?

### ISSUE THREAD

**Issue by polarathene** [jimporter/mike#44]
Presently I have the build output from `mkdocs build` deployed without using `mkdocs deploy`. I'm not sure if I can keep this workflow if adopting `mike` to add in versioned docs support though?

The present Github Actions workflow for deployment looks like this:

```yml
name: 'Documentation'

on:
  workflow_dispatch:
  push:
    branches:
      - master
    paths:
      - '.github/workflows/deploy-docs.yml'
      - 'docs/**'

# Jobs will run shell commands from this subdirectory:
defaults:
  run:
    working-directory: docs

jobs:
  deploy:
    name: 'Deploy Docs'
    runs-on: ubuntu-20.04
    steps:
      - uses: actions/checkout@v2

      - name: 'Build with mkdocs-material via Docker'
        run: docker run --rm -v ${PWD}:/docs squidfunk/mkdocs-material build --strict

      - name: 'Deploy to Github Pages'
        uses: peaceiris/actions-gh-pages@v3
        with:
          github_token: ${{ secrets.GITHUB_TOKEN }}
          publish_dir: ./docs/site
          user_name: 'github-actions[bot]'
          user_email: '41898282+github-actions[bot]@users.noreply.github.com'
```

- The `peaceiris/actions-gh-pages@v3` action is preferred for generic deploys of build output.
- I'm using the docker container for `mkdocs-material` which simplifies maintenance, but it seems to use `mike` I'll need to include additional environment setup to add in python and pip install packages?

My understanding is that the main project docs remain as is, but `mike` will be managing versions with subdirectories of versions/aliases on the `gh-pages` branch, that it's using this branch to add the new build output to the configured location, and pushing that new commit for deployment?

Other `mike` features aside, does that mean as long as I mirror `gh-pages` in `publish_dir` location in my deployment action, and add my build to the correct version subdirectory, I'd have the equivalent of `mike deploy`?

**Comment by jimporter** [comment]
Yeah, if you're not using `mike deploy`, then you don't really gain much from using mike to begin with. The `deploy` subcommand does do a few secondary things like injecting a version selector on the built-in MkDocs themes and fixing the `site_dir` value in `mkdocs.yml`, but other than that all it's doing is taking the results of `mkdocs build`, putting them in a subdirectory of `gh-pages`, and updating `versions.json` as appropriate.

That said, I took a look at `peaceiris/actions-gh-pages` and couldn't see anything it did that `mike deploy` can't, so I don't think you'd actually lose out on any features if you switched away from that to using mike instead. For the Docker side of things, it might be worth filing an issue at the mkdocs-material repo to suggest adding support for invoking mike. Since mkdocs-material supports mike now, it'd make sense for the Docker image to support it too.

**Comment by polarathene** [comment]
> For the Docker side of things, it might be worth filing an issue at the mkdocs-material repo to suggest adding support for invoking mike.

I try to avoid collaborating with the maintainer there now.

They've been difficult to work with personally, I some how within a brief period racked up two violation warnings against their Code of Conduct (1st due to a miscommunication, 2nd due to describing expectations of a fix I provided being upstreamed under the maintainership style).

Any interaction I have there is often dismissed, I doubt requesting such would be welcomed (at least when coming from me).

---

> That said, I took a look at `peaceiris/actions-gh-pages` and couldn't see anything it did that `mike deploy` can't, so I don't think you'd actually lose out on any features if you switched away from that to using mike instead.

Yeah, that wasn't a concern. It's just the preferred action for workflows I manage in general, last I recall `mkdocs deploy` required more config in the workflow while also being more implicit about a few things like how auth was configured to push to the `gh-pages` branch (as it was using `run` shell commands it accessed the available env vars IIRC, whereas here I pass the secret and would need to do similar for `docker run`).

I think both `mkdocs-material` ([Publishing site docs](https://squidfunk.github.io/mkdocs-material/publishing-your-site/)) and `mike` advocate deploying in the same manner ([as does MkDocs itself](https://www.mkdocs.org/user-guide/deploying-your-docs/)) without using Docker or other Actions where more explicit config needs to be forwarded.

None really seem to detail much about the implicit settings picked up from environment, which put me off those a bit as I'm collaborating with another user with our own testing forks for contributing a docs migration to an organization that migrates from Github Wiki to MkDocs. With the deploy action, it was more flexible/portable (trialed it with other static doc generators prior to settling on MkDocs, each of those had their own `deploy` commands and expectations as well).

---

When asked to look into versioning for MkDocs, I was aware of `mike` and have been trying to grok that approach. `mike` looks quite neat for managing this too with it's various features. I'm mostly interested in retaining the current deploy action and having mike slot in prior to do it's thing.

I haven't looked through your code yet, but:

- Is `mike deploy` the only method available that sorts out the `gh-pages` layout? (version and alias subdirectories, redirects, etc)
- Is there a local copy made of the `gh-pages` branch, or a switch to that branch that adds the new build output to commit & push during `mike deploy`?
- Could it perform all that without the push, where my deploy action just references the directory `mike` prepped? Or is it likely to be incompatible with how `mike` approaches the update?
- Are other `mike` commands directly making modifications to `gh-pages` similar to `mike deploy`?

---

> Yeah, if you're not using `mike deploy`, then you don't really gain much from using mike to begin with.

Mostly interested in using the drop-down version select, which afaik references a json config from `mike` to map to directories of each version/alias prefix to go to (with alias using redirects that `mike` generates instead of a copy of an existing version build?).

With my current action there seems to be a [deploy to subdirectory](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-deploy-to-subdirectory-destination_dir) feature. Is this what `mike deploy` is roughly doing in regards to updating the current built docs output to `gh-pages` branch? (other version and alias directories being handled via other `mike` commands?)

Do your users usually run the other `mike` commands locally such as when a new major version is released, or are they also handling that via CI automation?

**Comment by jimporter** [comment]
> Yeah, that wasn't a concern. It's just the preferred action for workflows I manage in general, last I recall `mkdocs deploy` required more config in the workflow while also being more implicit about a few things like how auth was configured to push to the `gh-pages` branch (as it was using `run` shell commands it accessed the available env vars IIRC, whereas here I pass the secret and would need to do similar for `docker run`).

Yeah, there are a couple of complexities to pushing to `gh-pages` from an action; I think you need to set up auth (sometimes) and also `fetch-depth: 0` on the checkout so that you actually get a checkout that includes the `gh-pages` branch (see [here](https://github.com/jimporter/mike/issues/28#issuecomment-730813808)). There's probably some value in having a `mike deploy` GH action, but it's not been a priority for me since mike already suits all of my needs as-is.

> * Is `mike deploy` the only method available that sorts out the `gh-pages` layout? (version and alias subdirectories, redirects, etc)

`mike alias` and `mike delete` also manipulate these things, but they're not much different from `mike deploy`.

> * Is there a local copy made of the `gh-pages` branch, or a switch to that branch that adds the new build output to commit & push during `mike deploy`?

Yes, mike always operates on a local checkout of the `gh-pages` branch.

> * Could it perform all that without the push, where my deploy action just references the directory `mike` prepped? Or is it likely to be incompatible with how `mike` approaches the update?

You don't have to push the commit from `mike deploy` if you don't want to. So long as you're ok with handling a few of the bits yourself, you can really do whatever you want with the local commit from `mike deploy`.
 
> * Are other `mike` commands directly making modifications to `gh-pages` similar to `mike deploy`?

Every mike subcommand that modifies things (so, all of them except `list` and `serve`) modify your local checkout of `gh-pages`.

> Mostly interested in using the drop-down version select, which afaik references a json config from `mike` to map to directories of each version/alias prefix to go to (with alias using redirects that `mike` generates instead of a copy of an existing version build?).

In theory, if you maintain `versions.json` yourself and put your docs in a `/[version]/` subdirectory (and handle `site_url` if you use it), you should be able to avoid using mike entirely. I think the Material theme would be able to handle this, and the built-in themes would work if you copied the version-selector code from [mike/themes](https://github.com/jimporter/mike/tree/master/mike/themes).

In fact, I'd be happy to see other projects use the `versions.json` + subdirectory approach to build docs, since people might want to manage doc deployment differently.

> With my current action there seems to be a [deploy to subdirectory](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-deploy-to-subdirectory-destination_dir) feature. Is this what `mike deploy` is roughly doing in regards to updating the current built docs output to `gh-pages` branch? (other version and alias directories being handled via other `mike` commands?)

Essentially, yes. It also updates `versions.json` if needed.

> Do your users usually run the other `mike` commands locally such as when a new major version is released, or are they also handling that via CI automation?

I suppose it probably varies from user to user. My *intent* with mike was that users would write automation scripts around it to supply the project's current version to mike so that you just run some command like `python build_docs.py` and things Just Work. Since project versions are specified in many different ways, that's something mike doesn't handle itself.

**Comment by polarathene** [comment]
> There's probably some value in having a `mike deploy` GH action, but it's not been a priority for me since mike already suits all of my needs as-is.

Yeah no worries there :smile: 

> `mike alias` and `mike delete` also manipulate these things, but they're not much different from `mike deploy`.

> You don't have to push the commit from `mike deploy` if you don't want to. So long as you're ok with handling a few of the bits yourself, you can really do whatever you want with the local commit from `mike deploy`.

> Every mike subcommand that modifies things (so, all of them except `list` and `serve`) modify your local checkout of `gh-pages`.

Ahhh... I think I must have misunderstood your docs at some point as I was under the impression `mike deploy` was implicitly pushing to the origin/remote `gh-pages` branch. 

---

With redirects/aliases, this wouldn't be something required if `latest` represented the `master` branch version of docs, and any tagged releases create a subdirectory matching `<major>.<minor>` listed in `versions.json`?

I'm new to versioning docs, so perhaps you're aware of a case where redirects and aliases might be useful with this basic versioning approach? (I saw the case for `.<patch>` aliases, but could list these as `.x` since no content should change in a meaningful way)

> In fact, I'd be happy to see other projects use the `versions.json` + subdirectory approach to build docs, since people might want to manage doc deployment differently.

Was this referring to other version management projects like Traefik's [`structor`](https://github.com/traefik/structor) that seems to build all semver tags/branches? The building of previous versions seems a bit unnecessary if no changes were made (especially since it appears to build docs as they were back then regarding styles and plugins), but they seem to inject JS/CSS that adds a version list into the sidebar instead; your version selector approach using a `versions.json` seems nicer :)

> Since project versions are specified in many different ways, that's something mike doesn't handle itself.

Oh, I didn't mean that. Just if users were running other `mike` commands locally instead of via CI on events. Often the CI can provide that metadata like version tags.

Now that it's more clear that `mike` is performing these actions on a local copy of `gh-pages` and that the user must still push that to the remote themselves (or through automation) as a separate command,  this question was probably a bit silly :sweat_smile: 

> and handle `site_url` if you use it

Is this just modifying it in `mkdocs.yml` before building docs, or is it a post-process performed on build output? It seems to be required to be specified if `sitemap.xml` is to generate anything useful (otherwise all it's values are empty).

---

Right, well I'll see how it goes. Thanks for creating `mike` and for taking the time to respond answering my questions here :)

**Comment by jimporter** [comment]
> Ahhh... I think I must have misunderstood your docs at some point as I was under the impression `mike deploy` was implicitly pushing to the origin/remote `gh-pages` branch.

You can explicitly push by passing `-p`/`--push`, but by default it doesn't do that.

> With redirects/aliases, this wouldn't be something required if `latest` represented the `master` branch version of docs, and any tagged releases create a subdirectory matching `<major>.<minor>` listed in `versions.json`?

Yeah, if every subdirectory is a different version, aliases aren't necessary.

> I'm new to versioning docs, so perhaps you're aware of a case where redirects and aliases might be useful with this basic versioning approach? (I saw the case for `.<patch>` aliases, but could list these as `.x` since no content should change in a meaningful way)

What I do (and what mike's feature set is built to support) is that every version of the docs lives at `/x.y/`, the latest *released* version has an alias of `/latest/` and the current development version is aliased with `/dev/`. So if you just published 0.1.2 and you're working on 0.2.0, you'd have the following:

```
/0.1/    => docs for 0.1.2
/0.2/    => docs for 0.2.0-dev
/latest/ => aliased to /0.1/
/dev/    => aliased to /0.2/
```

> Was this referring to other version management projects like Traefik's [`structor`](https://github.com/traefik/structor) that seems to build all semver tags/branches? The building of previous versions seems a bit unnecessary if no changes were made (especially since it appears to build docs as they were back then regarding styles and plugins), but they seem to inject JS/CSS that adds a version list into the sidebar instead; your version selector approach using a `versions.json` seems nicer :)

I wasn't specifically thinking of them, but for example, someone out there might want to do something similar to mike, but with support for a real server like Apache. In that case, you'd be able to do aliases via `mod_rewrite`, and could even do some server-side logic to render the version selector if you wanted. In a case like that, it'd be cool if they used `versions.json` so that it's easy to migrate from mike to this new hypothetical tool.

It'd also be cool if other non-MkDocs projects could understand `versions.json` so you could switch to/from, say, Sphinx.

> Oh, I didn't mean that. Just if users were running other `mike` commands locally instead of via CI on events. Often the CI can provide that metadata like version tags.

For what it's worth, I usually run mike locally, but it's mostly because I haven't bothered to set up docs generation on CI. I usually double-check my generated docs manually before pushing them, so doing it automatically doesn't save me much effort.

> Is this just modifying it in `mkdocs.yml` before building docs, or is it a post-process performed on build output? It seems to be required to be specified if `sitemap.xml` is to generate anything useful (otherwise all it's values are empty).

The former, essentially. The `mike` MkDocs plugin (which is automatically injected into `mkdocs.yml` when you call `mike deploy`, but you can specify it explicitly to configure it) adds the current version number to the `site_url`, so if it's originally `https://example.com/`, and you call `mike deploy 1.0`, the `site_url` becomes `https://example.com/1.0`. That way the `<link rel="canonical">` tag points to the right place.

> Right, well I'll see how it goes. Thanks for creating `mike` and for taking the time to respond answering my questions here :)

Good luck! Hopefully everything works ok; I've tried to make it reasonably flexible, but everyone has different requirements so there might be some bumps. :)

**Comment by polarathene** [comment]
> So if you just published 0.1.2 and you're working on 0.2.0, you'd have the following

Yeah that's a nice use case :smiley: 

I'm opting to follow how `mike` handles `gh-pages` , and if the maintainers are ok with adopting `mike` later to have a nicer way to manage versions it should be a pretty seamless transition.

> someone out there might want to do something similar to mike, but with support for a real server like Apache.

Ah yep, I just setup a local server to test the built docs under version subpaths with a `versions.json` text and got it working. So someone would make a third-party add-on that'd read the `versions.json` and automatically handles the routing and aliases, that'd be pretty cool! :+1: 

> It'd also be cool if other non-MkDocs projects could understand versions.json so you could switch to/from, say, Sphinx.

I think that'd be cool too, but the switch may not always be as easy if depending on a bunch of plugin/extensions tied to MkDocs. 

I know the docs I'm working with are leveraging non-standard markdown syntax. Still, not having to worry about version routing again would definitely be a nice perk! :+1: 

> I usually double-check my generated docs manually before pushing them, so doing it automatically doesn't save me much effort.

We'll be having automated preview deploys on Vercel for PRs for that, so when those are merged to master branch, CI will handle the production deploy :)

> adds the current version number to the `site_url`, so if it's originally `https://example.com/`, and you call `mike deploy 1.0`, the `site_url` becomes `https://example.com/1.0`. That way the `<link rel="canonical">` tag points to the right place.

Ah, good to know. Traefik seems to be approaching this with their rather new (Feb 2021) tool [`seo-doc`](https://github.com/traefik/seo-doc) which updates the `<link rel="canonical">` to point to the latest docs (which they host without a subdirectory, while all others are version tag subdirectories or an "experimental" release branch), they also seem to strip out `sitemap.xml` from the older docs and discourage search engines from crawling them AFAIK from going over that README.

> Good luck! Hopefully everything works ok; I've tried to make it reasonably flexible, but everyone has different requirements so there might be some bumps. :)

Thanks! I can see plenty of thought gone into `mike`, and I like that we should be able to change the CI process to use it if the maintainers feel it's worth the extra dependencies (which they're trying to keep to a minimum, so if `mkdocs-material` does eventually bundle `mike` that'd probably help).
